### PR TITLE
MODORDERS-154 PO Line: make location repeatable

### PR DESCRIPTION
**Pull request by piotr-kalashuk**
## Purpose
[MODORDERS-154](https://issues.folio.org/browse/MODORDERS-154) PO Line: make location details repeatable

## Approach
- Switching to schemas where location is repeatable
- Update logic to work with list of locations
- Inventory interaction
  - The holding is created for each unique location of the particular PO Line
  - The items are created taking into account corresponding holdings
- Update all mock files
- Validation
  - physical/electronic quantity should depend on PO Line's `order_format`
  - sum of physical quantity of all locations must not exceed `cost.quantity_physical`
  - sum of electronic quantity of all locations must not exceed `cost.quantity_electronic`

### TODO
- [x] Validation of the PO Line
- [ ] Validation of the PO Line's locations presence cannot be done until UIOR-121 implemented

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards ([sonar mod-orders PR-100](https://sonarcloud.io/project/issues?id=org.folio%3Amod-orders&pullRequest=100&resolved=false))?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [x] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/rest/impl/InventoryHelper.java [hunk 1]
```diff
@@ -34,13 +34,34 @@
...
+  static final String INSTANCE_IDENTIFIER_TYPE_VALUE = "value";
+  static final String HOLDING_INSTANCE_ID = "instanceId";
+  static final String HOLDING_PERMANENT_LOCATION_ID = "permanentLocationId";
+  static final String ITEM_HOLDINGS_RECORD_ID = "holdingsRecordId";
+  static final String ITEM_STATUS = "status";
+  static final String ITEM_STATUS_NAME = "name";
+  static final String ITEM_STATUS_ON_ORDER = "On order";
+  static final String ITEM_MATERIAL_TYPE_ID = "materialTypeId";
+  static final String ITEM_PERMANENT_LOAN_TYPE_ID = "permanentLoanTypeId";
+  static final String ITEM_PURCHASE_ORDER_LINE_IDENTIFIER = "purchaseOrderLineIdentifier";
```
Comment (craigmcnally) [inline src/main/java/org/folio/rest/impl/InventoryHelper.java]: Unless there's a reason for this, can you please make these private so their consistent with the other constants below?

File: src/main/java/org/folio/rest/impl/InventoryHelper.java [hunk 2]
```diff
@@ -78,25 +101,68 @@
...
+    return collectResultsOnSuccess(itemsPerHolding);
+  }
+
+  /**
+   * Wait for all requests completion and collect all resulting objects In case any failed, complete resulting future with the exception
+   * @param futures list of futures and each produces list of resulting objects on completion
+   * @param <T> resulting type
+   * @return resulting objects
+   */
+  private <T> CompletableFuture<List<T>> collectResultsOnSuccess(List<CompletableFuture<List<T>>> futures) {
```
Comment (craigmcnally) [inline src/main/java/org/folio/rest/impl/InventoryHelper.java]: In general I like the idea of pulling out and generalizing this code as it's a common pattern used throughout the class (and module).  Unfortunately, it seems that minor variations prevent this from being used in at least one other case in this class.  I wonder if that indicates that you've stopped just short of making this truly useful, or if we should consider moving this to somewhere in the code which allows it to be easily called from other parts of the code.

See lines 399 - 404

Comment (craigmcnally) [reply src/main/java/org/folio/rest/impl/InventoryHelper.java]: Just food for thought... this isn't something that should prevent this from being merged.

File: src/main/java/org/folio/rest/impl/InventoryHelper.java [hunk 3]
```diff
@@ -78,25 +101,68 @@
...
+    return VertxCompletableFuture.allOf(futures.toArray(new CompletableFuture[0]))
+      .thenApply(v -> futures
+        .stream()
+        // The CompletableFuture::join can be used safely because the `allOf` is used
```
Comment (craigmcnally) [inline src/main/java/org/folio/rest/impl/InventoryHelper.java]: I appreciate this comment.

File: src/main/java/org/folio/rest/impl/OrdersImpl.java [hunk 4]
```diff
@@ -32,7 +37,7 @@
 
   private static final Logger logger = LoggerFactory.getLogger(OrdersImpl.class);
 
-  private static final String ORDERS_LOCATION_PREFIX = "/orders/composite-orders";
+  private static final String ORDERS_LOCATION_PREFIX = "/orders/composite-orders/%s";
   private static final String ORDER_LINE_LOCATION_PREFIX = "/orders/order-lines/%s";
   public static final String OVER_LIMIT_ERROR_MESSAGE = "Your FOLIO system is configured to limit the number of PO Lines on each order to %s.";
```
Comment (craigmcnally) [inline src/main/java/org/folio/rest/impl/OrdersImpl.java]: Shouldn't these be moved to ErrorCodes?

File: src/main/java/org/folio/rest/impl/InventoryHelper.java (hunk 1 above)
Comment (piotr-kalashuk) [reply src/main/java/org/folio/rest/impl/InventoryHelper.java]: I decided to define Inventory objects' properties as constants just to use them also in unit tests to verify expected content of the inventory records. This is why all of them are package private.

File: src/main/java/org/folio/rest/impl/InventoryHelper.java (hunk 2 above)
Comment (piotr-kalashuk) [reply src/main/java/org/folio/rest/impl/InventoryHelper.java]: I moved `collectResultsOnSuccess` to `HelperUtils` and updated other classes to use it.

**Pull request merged**